datumaro/plugins/transforms.py

```diff
@@ -564,6 +567,66 @@ def transform_item(self, item):
                 annotations.append(ann.wrap())
         return item.wrap(annotations=annotations)
 
+class ProjectLabels(ItemTransform):
+    """
+    Changes the order of labels in the dataset from the existing
+    to the desired one and removes unknown labels. Updates or removes
+    the corresponding annotations.
+
+    Labels are matched by names (case dependent).
+
+    Useful for merging similar datasets.
+    """
+
+    def __init__(self, extractor: IExtractor, dst_labels: LabelCategories):
+        super().__init__(extractor)
+
+        self._categories = {}
+
+        src_label_cat = self._extractor.categories().get(AnnotationType.label)
+        self._make_label_id_map(src_label_cat, dst_labels)
+
+        src_mask_cat = self._extractor.categories().get(AnnotationType.mask)
+        if src_mask_cat is not None:
+            assert src_label_cat is not None
+            dst_mask_cat = MaskCategories(attributes=src_mask_cat.attributes)
+            dst_mask_cat.colormap = {
+                id: src_mask_cat[id]
+                for id, _ in enumerate(src_label_cat.items)
+                if id in src_mask_cat and (self._map_id(id) or id == 0)
+            }
+            self._categories[AnnotationType.mask] = dst_mask_cat
+
+        src_point_cat = self._extractor.categories().get(AnnotationType.points)
+        if src_point_cat is not None:
+            assert src_label_cat is not None
+            dst_point_cat = PointsCategories(attributes=src_point_cat.attributes)
+            dst_point_cat.items = {
+                id: src_point_cat[id]
+                for id, _ in enumerate(src_label_cat.items)
+                if id in src_point_cat and (self._map_id(id) or id == 0)
+            }
+            self._categories[AnnotationType.points] = dst_point_cat
+
```

Comment:
IMO, there should be a check here that no categories of unknown types are present. Otherwise, if a new category type is added in the future and this class is not updated, then old label IDs could leak into the transformed dataset.

Reply:
> there should be a check

Probably, a good point.

> could leak

Input categories are not copied in these transforms, so no.

Reply:
> Input categories are not copied in these transforms, so no.

Okay, fair point.

However, the annotations _are_ copied. So perhaps we should make sure to only copy annotations of known types so that we can make sure that annotations aren't copied without the appropriate categories?

Reply:
Currently, addition of new annotation type is a rare event, and addition of a new categories is even more rare. I suppose, when it happen, it will require a number of similar changes across the code.